src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,179 @@
+          if this function returns false, this value will remain unchanged.
+        :return bool if some value has been dropped (and updated as 
+          droppedValue) 
+      */
+      bool append(T newValue, T* droppedValue) {
+        //only in this case we drop oldest; this happens always after
+        //first maxCap steps ; must be checked before append()
+        const bool isFull = (buffer_.size()==maxCapacity);
+        if(isFull) {
+          *droppedValue = buffer_[idxNext_];
+        }
+        append(newValue);
+        return isFull;
+      }
+
+
+      /**
+        :return unordered content (data ) of this sl. window; 
+          call getLinearizedData() if you need them oredered from 
+          oldest->newest
+        This direct access method is fast.
+      */
+      const std::vector<T>& getData() const {
+        return buffer_;
+      }
+
+
+      /** linearize method for the internal buffer; this is slower than 
+        the pure getData() but ensures that the data are ordered (oldest at
+        the beginning, newest at the end of the vector
+        This handles case of |5,6;1,2,3,4| => |1,2,3,4,5,6|
+        :return new linearized vector
+      */
+      std::vector<T> getLinearizedData() const {
+        std::vector<T> lin;
+        lin.reserve(buffer_.size());
+
+        //insert the "older" part at the beginning
+        lin.insert(std::begin(lin), std::begin(buffer_) + idxNext_, std::end(buffer_));
+        //append the "newer" part to the end of the constructed vect
+        lin.insert(std::end(lin), std::begin(buffer_), std::begin(buffer_) + idxNext_);
+        return lin;
+      }
+
+
+      bool operator==(const SlidingWindow& r2) const {
+        const bool sameSizes = (this->size() == r2.size()) && (this->maxCapacity == r2.maxCapacity);
+        if(!sameSizes) return false; 
+        const std::vector<T> v1 = this->getLinearizedData(); 
+        const std::vector<T> v2 = r2.getLinearizedData();
+        return sameSizes && std::equal(v1.cbegin(), v1.cend(), v2.cbegin()); //also content must be same
+      }
+
+
+      bool operator!=(const SlidingWindow& r2) const {
+        return !operator==(r2);
+      }
+
+
+      /** operator[] provides fast access to the elements indexed relatively
+        to the oldest element. So slidingWindow[0] returns oldest element, 
+        slidingWindow[size()] returns the newest. 
+      :param UInt index - index/offset from the oldest element, values 0..size()
+      :return T - i-th oldest value in the buffer
+      :throws 0<=index<=size()
+      */ 
+      T& operator[](UInt index) {
+        NTA_ASSERT(index <= size());
+        NTA_ASSERT(size() > 0);
+        //get last updated position, "current"+index(offset)
+        //avoid calling getLinearizeData() as it involves copy()
+        if (size() == maxCapacity) {
+          return &buffer_[(idxNext_ + index) % maxCapacity];
+        } else {
+          return &buffer_[index];
```

Comment:
@mrcslws I'm having troubles with this code. 
I think it used to work, as you've suggested https://github.com/numenta/nupic.core/pull/1277#discussion_r117592239 
Now I'm getting 
```
[ 78%] Building CXX object src/CMakeFiles/unit_tests.dir/test/unit/utils/SlidingWindowTest.cpp.o
In file included from /home/mmm/devel/HTM/nupic.core/src/test/unit/utils/SlidingWindowTest.cpp:24:
/home/mmm/devel/HTM/nupic.core/src/nupic/utils/SlidingWindow.hpp:158:18: error: non-const lvalue reference to type 'int' cannot bind to a temporary of type
      'value_type *' (aka 'int *')
          return &buffer_[(idxNext_ + index) % maxCapacity];
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/mmm/devel/HTM/nupic.core/src/test/unit/utils/SlidingWindowTest.cpp:46:16: note: in instantiation of member function
      'nupic::util::SlidingWindow<int>::operator[]' requested here
    ASSERT_EQ(w[1], w2[0]); //==1
               ^
```
And I'm really not sure what to do with that. Can you please advise?